File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/VersatileWeaponTraitEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum VersatileWeaponTraitEnum implements WeaponTraitEnum {
    VERSATILE_S(String.format("versatile_%s", PhysicalDamageTypeEnum.SLASHING)),
    VERSATILE_P(String.format("versatile_%s", PhysicalDamageTypeEnum.PIERCING)),
    VERSATILE_B(String.format("versatile_%s", PhysicalDamageTypeEnum.BLUDGEONING));

    private final String value;
    private final static Map<String, VersatileWeaponTraitEnum> CONSTANTS = new HashMap<>();

    static {
        for (VersatileWeaponTraitEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    VersatileWeaponTraitEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static VersatileWeaponTraitEnum fromValue(String value) {
        VersatileWeaponTraitEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/VersatileWeaponTraitEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class VersatileWeaponTraitEnum(private val value: String) : WeaponTraitEnum {
    VERSATILE_S(String.format("versatile_%s", PhysicalDamageTypeEnum.SLASHING)), VERSATILE_P(String.format("versatile_%s", PhysicalDamageTypeEnum.PIERCING)), VERSATILE_B(String.format("versatile_%s", PhysicalDamageTypeEnum.BLUDGEONING));

    companion object {
        private val CONSTANTS: MutableMap<String, VersatileWeaponTraitEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): VersatileWeaponTraitEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}